File: main/src/com/seovic/coherence/loader/DefaultLoader.java
package com.seovic.coherence.loader;


import com.seovic.core.Extractor;
import com.seovic.core.Updater;


/**
 * Default loader implementation.
 *
 * @author Aleksandar Seovic/Ivan Cikic  2009.06.15
 */
public class DefaultLoader
        implements Loader
    {
    // ---- constructors ----------------------------------------------------

    /**
     * Construct DefaultLoader instance.
     *
     * @param source  source to load items from
     * @param target  target to load items into
     */
    public DefaultLoader(Source source, Target target)
        {
        this.source = source;
        this.target = target;
        }


    // ---- Loader implementation -------------------------------------------

    /**
     * {@inheritDoc}
     */
    public void load()
        {
        source.beginExport();
        target.beginImport();
        String[] propertyNames = target.getPropertyNames();

        for (Object sourceItem : source)
            {
            Object targetItem = target.createTargetInstance(source, sourceItem);
            for (String property : propertyNames)
                {
                Extractor extractor = source.getExtractor(property);
                Updater   updater   = target.getUpdater(property);
                updater.update(targetItem, extractor.extract(sourceItem));
                }
            target.importItem(targetItem);
            }
        source.endExport();
        target.endImport();
        }

    // ---- data members ----------------------------------------------------

    /**
     * Source to load items from.
     */
    private Source source;

    /**
     * Target to load items into.
     */
    private Target target;
    }

File: trunk/main.net/src/Coherence.Tools/Loader/DefaultLoader.cs
using Seovic.Core;

namespace Seovic.Loader
{
    /// <summary>
    /// Default loader implementation.
    /// </summary>
    /// <author>Aleksandar Seovic  2009.06.15</author>
    /// <author>Ivan Cikic  2009.10.01</author>
    public class DefaultLoader : ILoader
    {
        #region Constructors

        /// <summary>
        /// Construct DefaultLoader instance.
        /// </summary>
        /// <param name="source">Source to load items from</param>
        /// <param name="target">Target to load items into</param>
        public DefaultLoader(ISource source, ITarget target)
        {
            this.source = source;
            this.target = target;
        }

        #endregion

        #region ILoader implementation

        /// <summary>
        /// Load items from the ISource into the ITarget.
        /// </summary>
        public void Load()
        {
            source.BeginExport();
            target.BeginImport();
            string[] propertyNames = target.PropertyNames;

            foreach (object sourceItem in source)
            {
                object targetItem = target.CreateTargetInstance(source, sourceItem);
                foreach (string property in propertyNames)
                {
                    IExtractor extractor = source.GetExtractor(property);
                    IUpdater updater     = target.GetUpdater(property);
                    updater.Update(targetItem, extractor.Extract(sourceItem));
                }
                target.ImportItem(targetItem);
            }
            source.EndExport();
            target.EndImport();
        }

        #endregion

        #region Data members

        /// <summary>
        /// Source to load items from.
        /// </summary>
        private ISource source;

        /// <summary>
        /// Target to load items into.
        /// </summary>
        private ITarget target;

        #endregion
    }
}
